api/timeseries_storage_api.go

```diff
@@ -21,6 +21,7 @@ import (
 )
 
 type TimeseriesStorageAPI interface {
+	AdjustTimerange(requested Timerange, slotLimit int) Timerange
```

Comment:
Right now I think the method name is certainly misleading. Having said that, it's not clear where else it should go- because the only place in the code that can know the internals of the API is the API implementation itself.

The specification for `AdjustTimerange` has the following requirements:

The resulting `Timerange` should:

1. Have (approximately) the same `Start` and `End` as the input (subject to changes due to resolution snapping)
2. Have a resolution *which the backend supports*
3. Have fewer slots than the specified `slotLimit`

The purpose of (1) is so that the user isn't shown a timerange that's completely different from what they asked for.

The purpose of (2) is so that the data can be queried without requiring up- or down-sampling. Otherwise, when the time to fetch the data occurs, the backend will find that it *needs* to sample the data at a new resolution.

The purpose of (3) is because if the result has too many slots, the query is rejected immediately, and nothing is evaluated (the user just gets back an error message).

The problem is essentially that (2) requires knowledge of the supported resolutions of the API, which only the API itself knows. Moreover, a list of resolutions isn't enough, because based on their TTLs or other characteristics, a given resolution might not be available for the specified `Timerange`.

One possibility would be instead of asking for a `Timerange`, just ask for a `resolution` (either in milliseconds or a `time.Duration`). Then the generic re-snapping can be performed at the call site, instead of inside the implementation.